CLN: c-lightning-Rest & RTL configuration (#1025)
raspibolt/raspibolt: guide/bonus/lightning/cln.md

  gpg:          There is no indication that the signature belongs to the owner.
  Primary key fingerprint: 3E9B D443 6C28 8039 CA82  7A92 00C9 E2BC 2E45 666F
  ```
  
* Extract and install with `npm`:

  ```sh
  $ tar xvf v0.9.0.tar.gz 
  $ cd c-lightning-REST-0.9.0
  $ npm install
  ```
  
* Copy content to plugin datadir:

  ```sh
  $ cp -r ~/c-lightning-REST-0.9.0/ /data/lightningd-plugins-available/
  ```

* Setup c-lightning-Rest as plugin:

  ```sh
  $ sudo nano /data/lightningd/config
  ```
  
* Add at the end of the file:

  ```ini
  # cln-rest-plugin
  plugin=/data/lightningd-plugins-available/c-lightning-REST-0.9.0/plugin.js
  rest-port=3092
  rest-docport=4091
  rest-protocol=http
  ```
  
* Add a sample config file and run the plugin once manually to create required `access.macaroon`
  
  ```sh
  $ cd /data/lightningd-plugins-available/c-lightning-REST-0.9.0
  $ cp sample-cl-rest-config.json cl-rest-config.json
  $ node cl-rest.js
  ```
  
* Now you should see a new folder being created in `/data/lightningd-plugins-available/c-lightning-REST-0.9.0/` called `certs` which contains the required `access.macaroon` for the next steps. Stop `cl-rest.js` by hitting CTRL+C.
  

> * As user `rtl`:

Reviewer: ```suggestion
* Go back to the "admin" user:
```

The next commands use sudo so should be invoked by admin

  ```sh
  $ exit

>   $ sudo su - rtl

Reviewer: ```suggestion
```

  ``` 

* Copy `access.macaroon` to home directory of user `rtl`:  
  
  ```sh
  $ sudo cp /data/lightningd-plugins-available/c-lightning-REST-0.9.0/certs/access.macaroon /home/rtl/
  $ sudo chown rtl:rtl /home/rtl/access.macaroon
  ```
  
* Restart `lightningd.service` and look for errors in cln's log: `tail -f /data/lightningd/cln.log`. Positive results look like this:

  ```
  UNUSUAL plugin-plugin.js: --- Starting the cl-rest server ---
  UNUSUAL plugin-plugin.js: --- cl-rest api server is ready and listening on port: 3092 ---
  UNUSUAL plugin-plugin.js: --- cl-rest doc server is ready and listening on port: 4091 ---
  ```

### Configuring RTL

* By the following configuration we tell RTL to connect to our CLN node. Change to user `rtl`:

  ```sh
  $ sudo su - rtl
  $ nano /home/rtl/RTL/RTL-Config.json
  ```

* Edit or add the following content. Adjust `multiPass` to your desired login password, if not already set.

  ```ini
  {
    "multiPass": "<yourfancypassword>",
    "port": "3000",
    "SSO": {
      "rtlSSO": 0,
      "rtlCookiePath": "",
      "logoutRedirectLink": ""
    },
    "nodes": [
      {
        "index": 1,
        "lnNode": "CLN Node",

>         "lnImplementation": "CLT",

Reviewer: ```suggestion
        "lnImplementation": "CLN",
```

From [RTL github](https://github.com/Ride-The-Lightning/RTL/blob/master/.github/docs/Application_configurations.md): 

> Allowed values LND/CLN/ECL

        "Authentication": {
          "macaroonPath": "/home/rtl",
          "configPath": "/data/lightningd/config"
        },
        "Settings": {
          "userPersona": "OPERATOR",
          "themeMode": "DAY",
          "themeColor": "INDIGO",
          "fiatConversion": true,
          "currencyUnit": "EUR",
          "logLevel": "ERROR",
          "lnServerUrl": "http://127.0.0.1:3092",
          "enableOffers": true
        }
      }
    ],
    "defaultNodeIndex": 1
  }
  ```

* As user admin, startup RTL and connect via browser: 

  ```sh
  $ exit
  $ sudo systemctl start rtl
  ```

> * Access RTL via your local IP: `https://<your-local-ip>:4002`

Reviewer: ```suggestion
* Access RTL via your local IP: `https://<your-local-ip>:4001`
```

.. shouldn't it be 4001 for RTL? https://raspibolt.org/guide/lightning/web-app.html#firewall--reverse-proxy

  $ sudo systemctl start cln.service
  ```

## c-lightning-Rest & RTL

* c-lightning-Rest: REST APIs for c-lightning written with node.js and provided within the [RTL repository](https://github.com/Ride-The-Lightning/c-lightning-REST).
* In this chapter we are going to add c-lightning-Rest as CLN plugin and furthermore connect RTL for CLN node management. 
* The installation of RTL is already described [here](https://raspibolt.org/guide/lightning/web-app.html), so we are focusing on the configuration.

### c-lightning-Rest plugin

> * Setting up c-lightning-Rest as plugin for CLN requires us to be user "cln". So first we need to download and verify the c-lightning-Rest package:

Reviewer 2: This is not a strict requirement. You can create separate `cln-rest` user. And then `ln -s /data/cln /home/cln-rest/.lightning`. Although I haven't tested that with running c-lightning-REST as a CLN plugin, I run it as a systemd service instead.

Author: Ok, good point. I'll rephrase.

  $ sudo systemctl start cln.service
  ```

## c-lightning-Rest & RTL

* c-lightning-Rest: REST APIs for c-lightning written with node.js and provided within the [RTL repository](https://github.com/Ride-The-Lightning/c-lightning-REST).
* In this chapter we are going to add c-lightning-Rest as CLN plugin and furthermore connect RTL for CLN node management. 

> * The installation of RTL is already described [here](https://raspibolt.org/guide/lightning/web-app.html), so we are focusing on the configuration.

Reviewer: ```suggestion
* The installation of RTL is already described [here](../../lightning/web-app.md), so we are focusing on the configuration.
```

### c-lightning-Rest plugin

* Setting up c-lightning-Rest as plugin for CLN. First we need to download and verify the c-lightning-Rest package:

  ```sh
  $ sudo su - cln

>   $ cd ~

Reviewer: Just a nit, but isn't this redundant as when opening a new user session we land in the home directory f that user? Maybe i'm not thinking of some edge cases..

Author: We switch admin and cln back and forth in this guide a few times and not every user will set this up or wants to set this up at a later point. So I thought this is just a reminder to switch to user cln, in case one isn't already.

>   $ wget https://github.com/Ride-The-Lightning/c-lightning-REST/archive/refs/tags/v0.7.2.tar.gz
>   $ wget https://github.com/Ride-The-Lightning/c-lightning-REST/releases/download/v0.7.2/v0.7.2.tar.gz.asc

Reviewer: ```suggestion
  $ wget https://github.com/Ride-The-Lightning/c-lightning-REST/archive/refs/tags/v0.9.0.tar.gz
  $ wget https://github.com/Ride-The-Lightning/c-lightning-REST/releases/download/v0.9.0/v0.9.0.tar.gz.asc
```

just updating to latest release

  ```

* Verify the release:

  ```sh

>   $ gpg --verify v0.7.2.tar.gz.asc v0.7.2.tar.gz

Reviewer: ```suggestion
  $ gpg --verify v0.9.0.tar.gz.asc v0.9.0.tar.gz
```

  ```
  
* Output should be like:

  ```sh
  gpg: Signature made Mon 09 May 2022 06:37:59 PM EDT
  gpg:                using RSA key 3E9BD4436C288039CA827A9200C9E2BC2E45666F
  gpg: Good signature from "saubyk (added uid) <39208279+saubyk@users.noreply.github.com>" [unknown]
  gpg:                 aka "Suheb (approves) <39208279+saubyk@users.noreply.github.com>" [unknown]
  gpg:                 aka "Suheb <39208279+saubyk@users.noreply.github.com>" [unknown]
  gpg: WARNING: This key is not certified with a trusted signature!
  gpg:          There is no indication that the signature belongs to the owner.
  Primary key fingerprint: 3E9B D443 6C28 8039 CA82  7A92 00C9 E2BC 2E45 666F
  ```
  
* Extract and install with `npm`:

  ```sh

>   $ tar xvf v0.7.2.tar.gz 
>   $ cd c-lightning-REST-0.7.2

Reviewer: ```suggestion
  $ tar xvf v0.9.0.tar.gz 
  $ cd c-lightning-REST-0.9.0
```

  $ npm install --only=prod
  ```
  
* Copy content to plugin datadir:

  ```sh

>   $ cp -r ~/c-lightning-REST-0.7.2/ /data/cl-plugins-available/

Reviewer: ```suggestion
  $ cp -r ~/c-lightning-REST-0.9.0/ /data/cl-plugins-available/
```

  ```

* Setup c-lightning-Rest as plugin:

  ```sh
  $ sudo nano /data/cln/config
  ```
  
* Add at the end of the file:

  ```ini
  # cln-rest-plugin

>   plugin=/data/cl-plugins-available/c-lightning-REST-0.7.2/plugin.js

Reviewer: ```suggestion
  plugin=/data/cl-plugins-available/c-lightning-REST-0.9.0/plugin.js
```

  rest-port=3092
  rest-docport=4091
  rest-protocol=http
  ```
  
* As user admin, create and setup user `rtl`:

  ```sh
  $ exit
  $ sudo adduser --disabled-password --gecos "" rtl
  $ sudo usermod -aG cln rtl
  ``` 

* Copy `access.macaroon` to home directory of user `rtl`:  
  
  ```sh

>   $ sudo cp /data/cl-plugins-available/c-lightning-REST-0.7.2/certs/access.macaroon /home/rtl/

Reviewer: ```suggestion
  $ sudo cp /data/cl-plugins-available/c-lightning-REST-0.9.0/certs/access.macaroon /home/rtl/
```

  $ sudo chown rtl:rtl /home/rtl/access.macaroon
  ```
  
* Restart `cln.service` and look for errors in cln's log: `tail -f /data/cln/cln.log`. Positive results look like this:

>   ```sh

Reviewer: ```suggestion
  ```
```

Author: is this correct? no replacement for the deleted line?

Reviewer: No sorry! I believe it's only the sh that I wanted to delete, oopsy :)

  UNUSUAL plugin-plugin.js: --- cl-rest doc server is ready and listening on port: 4091 ---
  ```

### Configuring RTL

* By the following configuration we tell RTL to connect to our CLN node. Change to user `rtl`:

  ```sh
  $ sudo su - rtl
  $ nano /home/rtl/RTL/RTL-Config.json
  ```

* Edit or add the following content. Adjust `multiPass` to your desired login password, if not already set.

  ```ini
  {
    "multiPass": "<yourfancypassword>",
    "port": "3000",
    "SSO": {
      "rtlSSO": 0,
      "rtlCookiePath": "",
      "logoutRedirectLink": ""
    },
    "nodes": [
      {
        "index": 1,
        "lnNode": "CLN Node",
        "lnImplementation": "CLT",
        "Authentication": {
          "macaroonPath": "/home/rtl",
          "configPath": "/data/cln/config"
        },
        "Settings": {
          "userPersona": "OPERATOR",
          "themeMode": "DAY",
          "themeColor": "INDIGO",
          "fiatConversion": true,
          "currencyUnit": "EUR",
          "logLevel": "ERROR",
          "lnServerUrl": "http://127.0.0.1:3092",
          "enableOffers": true
        }
      }
    ],
    "defaultNodeIndex": 1
  }
  ```

> * As user admin, startup RTL and connect via browser`: 

Reviewer: ```suggestion
* As user admin, startup RTL and connect via browser: 
```
